### PR TITLE
Fix diff skipping lines

### LINKED ISSUE #13153: Diff skips lines starting with '++ ' or '-- '
- Gitea version (or commit ref): master, 1.13-rc1, 1.12.x and earlier
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL): Compare https://try.gitea.io/arandomer/pathological/commit/88612bf69b8b2035bf41ae4fdbab0ae399238d60 with
 https://try.gitea.io/arandomer/pathological/commit/88612bf69b8b2035bf41ae4fdbab0ae399238d60.diff
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description

First noted in https://codeberg.org/Codeberg/Community/issues/308

Gitea's ParsePatch code in services/gitdiff/gitdiff.go will skip any diff lines that start with `"+++ "` or `"--- "`.

### PR DESCRIPTION
**Pull request by zeripath**
ParsePatch previously just skipped all lines that start with "+++ " or "--- "
and makes no attempt to see these lines in context.

This PR rewrites ParsePatch to pay attention to context and position
within a patch, ensuring that --- and +++ are only skipped if
appropriate.

This PR also fixes several issues with incomplete files.

Fix https://codeberg.org/Codeberg/Community/issues/308
Fix #13153

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by techknowlogick — COMMENTED** [review]
First, thank you so much for this PR! I've been afraid of our diff parsing, and so I'm especially appreciative that you took this on. Second, would you hate me if I asked for tests, specifically ones for diffs with files that have lines that start with `++` and `--`?

**Comment by zeripath** [comment]
Oh sorry I've spent so much time making the current tests pass I forgot to add the real one

**Comment by zeripath** [comment]
Done

File: services/gitdiff/gitdiff.go [hunk 1]
```diff
@@ -441,91 +441,249 @@ func (diff *Diff) LoadComments(issue *models.Issue, currentUser *models.User) er
...
 
-		trimLine := strings.Trim(line, "+- ")
+	if err == io.EOF {
+		err = nil
+	}
 
-		if trimLine == models.LFSMetaFileIdentifier {
-			curFileLFSPrefix = true
+	// FIXME: There are numerous issues with this:
+	// - we might one to consider detecting encoding while parsing but...
```
Comment (zeripath) [inline services/gitdiff/gitdiff.go]:
```suggestion
	// - we might want to consider detecting encoding while parsing but...
```

File: services/gitdiff/gitdiff.go [hunk 2]
```diff
@@ -441,91 +441,249 @@ func (diff *Diff) LoadComments(issue *models.Issue, currentUser *models.User) er
...
+	}
+
+	input := bufio.NewReaderSize(reader, readerSize)
+	line, err := input.ReadString('\n')
+	if err != nil {
+		return diff, err
+	}
+parsingLoop:
+	for {
+		if err != nil {
```
Comment (techknowlogick) [inline services/gitdiff/gitdiff.go]: I've attempted to trace through this loop to search for an err that is previously unhandled such that this conditional would be needed. Although it is an unreasonable hour for me to be awake, let alone doing PR reviews, so I am likely missing something, and should return in the morning with fresh eyes.

Comment (zeripath) [reply services/gitdiff/gitdiff.go]: Sorry that's a hang over from earlier code and error handing. I've removed it now and should have improved handling in general.

**Comment by lunny** [comment]
How about to use the package https://github.com/bluekeyes/go-gitdiff ?

**Comment by zeripath** [comment]
That's not a backportable change and would require lots of other changes.

This isn't the last change I'm going to make to diff though and it's possible that that library could form part of it.

**Review by techknowlogick — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**